pkg/oci/push.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2025 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/opencontainers/image-spec/specs-go"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/content/file"
+	"oras.land/oras-go/v2/registry/remote"
+)
+
+// Platform represents an OS/Arch platform for an OCI artifact.
+type Platform struct {
+	OS   string
+	Arch string
+}
+
+// Artifact describes an artifact to be pushed to an OCI registry.
+type Artifact struct {
+	// MediaType is the media type of the artifact.
+	// e.g. "application/vnd.pipecd.piped.plugin"
+	MediaType string
+	// ArtifactType is the type of the artifact.
+	// e.g. "application/vnd.pipecd.piped.plugin+type"
+	ArtifactType string
+	// FilePaths maps platforms to file paths.
+	FilePaths map[Platform]string
+}
+
+// PushFilesToRegistry pushes files described by the artifact to the target OCI registry URL.
+// It supports options for insecure connections.
+func PushFilesToRegistry(ctx context.Context, workDir string, artifact *Artifact, targetURL string, opts ...PushOption) error {
```

Comment:
[ask] Where is this method for pushing plugin executed? I want to know where you plan to push the plugin if you have idea about it.

Reply:
I want to implement commands like `pipectl plugin push` or `pipectl push plugin` to push plugins' binary to OCI Registries.

Reply:
Thank you, I got your idea.